### DISCUSSION TITLE
Handling date and time formats

### DISCUSSION THREAD

**Discussion by bmaschio** [jolie/jolie#572] [Ideas (Features, Fixes, etc.)]
@jolie/developers 
#71

Hi all I was wandering what you thing  about the possibility of adding a Date native type, in my opinion it would be useful in handling DB interaction in particular in the case of not SQL DB .
What do you think

Balint

**Comment by fmontesi** [comment]
Hi all,

My main problem with this is that usually there is not a single Date type
that makes sense for integrating with all systems. Different systems have
different representations.

Also, what would the advantage be with defining, instead, a standard
compound Date type? For example:

type Date:void { .year:int .month:int .day:int }

Then we could also define other types like Time, DateTime, etc., and put
all of them in an include file in the standard library.
- Fabrizio

**Comment by bmaschio** [comment]
@jolie/developers 
Hi all
I do agree that is not easy to define a single date format. but it would be very useful. we could elevate the date format at port level so that we could work using different format 

outputPort port1 {
  location: somelocationPort
  Interfaces: someInterfaces
  prococol: sodep{
                           .dateFormat="ddMMYYYY"
                           .dateTimeFormat= "ddMMYYYYHH:MM:SS"
                           }
}

What you think?

**Comment by fmontesi** [comment]
Sorry, I missed this comment.

Yes, but my question remains unanswered:
What is the advantage of native date types compared to user-defined types?

I can see more advantages for the latter (user-defined): If we are in a system using different sub-fields, we can adjust in the program.

**Comment by bmaschio** [comment]
By elevating the Date format to a complex native type like

dateType : string {
    .day:int
    .month:int 
    .year:int
} 

we will have not to write date check error code each time.
If i would have a inputPort with the possibility of define the format then the caller could recive a dateFormat error with also a message say ("Dude your date format is wrong you should send me a ddMMYYYY")

**Comment by klag** [comment]
We could address data format issue for string fields by using regex in
native types.
What do you think?

2016-03-21 18:10 GMT+01:00 Balint Maschio notifications@github.com:

**Comment by fmontesi** [comment]
Shouldn't the interface then be a string equipped with a regular expression
(that of the date format), instead of making a special case?

Even better, if we could decompose a string in subnodes through a regexp,
this would solve a lot of problems, not only dates but also emails etc.

Here's an example:

type Date : string( "(day:\w\w)... Etc" )

And then day would be a subnode automatically created.

I'm not sure I would have this automatic subnode creation as a feature of
Jolie though, I'd like it better if we had a good syntax sugar to have a
courier process doing it.

**Comment by fmontesi** [comment]
Now that both refinement types and modules are in git mainline, we have much better tools to attack this problem. We could create a `datetime` module (like in Python) with the necessary types and regexes. I'm updating the issue title accordingly.
